src/validator/RecoveryPlugin.sol

```diff
@@ -0,0 +1,222 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import "./IValidator.sol";
+import "openzeppelin-contracts/contracts/utils/cryptography/EIP712.sol";
+import "src/utils/KernelHelper.sol";
+
+struct RecoveryPluginStorage {
+    address owner;
+}
+
+struct Guardian {
+    address guardian;
+    uint256 weight;
+    bool approved;
+}
+
+contract RecoveryPlugin is IKernelValidator {
+    event OwnerChanged(
+        address indexed kernel,
+        address indexed oldOwner,
+        address indexed newOwner
+    );
+    event GuardianAdded(
+        address indexed kernel,
+        Guardian[] guardians,
+        uint256 weight
+    );
+
+    mapping(address => RecoveryPluginStorage) public recoveryPluginStorage;
+    mapping(address => Guardian[]) public guardians;
+    mapping(address => uint256) public thresholdWeight;
+    mapping(address => uint256) public recoveryDelay;
+
+    function disable(bytes calldata) external override {
+        delete recoveryPluginStorage[msg.sender];
+    }
+
+    function enable(bytes calldata _data) external override {
+        //0x00 - to add guardians
+        //0x01 - to change owner
+        bytes memory mode = bytes(_data[0:1]);
+        if (keccak256(mode) == keccak256(hex"00")) {
+            bytes32 weightinbytes = bytes32(_data[1:33]);
+            uint256 weight = uint256(weightinbytes);
+            bytes calldata guardiandata = bytes(_data[33:]);
+            addGuardian(guardiandata, weight, 1 days);
+        } else if (keccak256(mode) == keccak256(hex"01")) {
+            bytes calldata recoverydata = bytes(_data[1:]);
+            address newOwner = address(bytes20(recoverydata[0:20]));
+            bytes32 hash = bytes32(recoverydata[20:52]);
+            bytes[] memory signatures = divideBytes(bytes(recoverydata[52:]));
+            initRecovery(newOwner, hash, signatures);
+        } else {
+            revert("Invalid mode");
+        }
+    }
+
+    function divideBytes(
+        bytes memory data
+    ) public pure returns (bytes[] memory) {
+        require(data.length % 65 == 0, "Data length must be a multiple of 65");
+
+        uint chunks = data.length / 65;
+
+        bytes[] memory dividedBytes = new bytes[](chunks);
+
+        for (uint i = 0; i < chunks; i++) {
+            bytes memory chunk = new bytes(65);
+            for (uint j = 0; j < 65; j++) {
+                chunk[j] = data[(i * 65) + j];
+            }
+            dividedBytes[i] = chunk;
+        }
+        return dividedBytes;
+    }
+
+    function changeOwner(address _newOwner) public {
+        address oldOwner = recoveryPluginStorage[msg.sender].owner;
+        recoveryPluginStorage[msg.sender].owner = _newOwner;
+        for (uint256 i = 0; i < guardians[msg.sender].length; i++) {
+            guardians[msg.sender][i].approved = false;
+        }
+        emit OwnerChanged(msg.sender, oldOwner, _newOwner);
+    }
+
+    function addGuardian(
```

Comment:
Since this function actually adds ALL guardians, it might be best to call it `setGuardians`.

Also I think `setGuardians` should actually reset all guardians which I don't think this function currently does.  I think it's better that way since otherwise if you accidentally added a wrong guardian, as far as I can tell there's no way to remove it.